### PR TITLE
Increase coverage of spack.relocate

### PR DESCRIPTION
**Pull request by alalazo**
This PR:

- [x] Adds docstrings to a few functions involved in ELF relocation
- [x] Adds unit test to them
- [x] Adds `patchelf` to Travis to trigger tests depending on it

Continuation of #15610

### CODE REVIEW

File: lib/spack/spack/relocate.py [hunk 1]
```diff
@@ -482,19 +511,25 @@ def replace(match):
...
 
 
-def replace_prefix_nullterm(path_name, old_dir, new_dir):
-    """
-    Attempt to replace old install prefix with new install prefix
-    in binary files by replacing with null terminated string
-    that is the same length unless the old path is shorter
-    Used on linux to replace mach-o rpaths
-    """
+def _replace_prefix_nullterm(filename, old_dir, new_dir):
```
Comment (alalazo) [inline lib/spack/spack/relocate.py]: @gartung I noticed that this function is not used at all in the code base. Is it a good candidate for removal or was it maintained on purpose for some reason?

File: lib/spack/spack/relocate.py [hunk 2]
```diff
@@ -385,57 +386,79 @@ def macholib_get_paths(cur_path):
...
+        bak_path = target + ".bak"
+        shutil.copy(target, bak_path)
 
+    patchelf, output = executable.Executable(bak_path or _patchelf()), None
     try:
-        patchelf('--force-rpath', '--set-rpath', '%s' % new_joined,
-                 '%s' % path_name, output=str, error=str)
+        # TODO: revisit the use of --force-rpath as it might be conditional
+        # TODO: if we want to support setting RUNPATH from binary packages
+        patchelf_args = ['--force-rpath', '--set-rpath', rpaths_str, target]
```
Comment (alalazo) [inline lib/spack/spack/relocate.py:416]: @gartung If I understand the code correctly this function will always set RPATHs and never set RUNPATHs regardless of what is in `config.yaml`? Is it correct?

File: lib/spack/spack/relocate.py (hunk 1 above)
Comment (gartung) [reply lib/spack/spack/relocate.py]: I used it to set rpaths in mach-o binaries when they are installed on a linux server.

File: lib/spack/spack/relocate.py (hunk 2 above)
Comment (gartung) [reply lib/spack/spack/relocate.py:416]: As far as I know patchelf only handles rpaths.

File: lib/spack/spack/relocate.py (hunk 1 above)
Comment (gartung) [reply lib/spack/spack/relocate.py]: But I see that I dropped that case at some point.

**Review by gartung — APPROVED** [review]

File: lib/spack/spack/relocate.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/relocate.py]: I'll remove it then. In case we need it later we can retrieve it from git history. Is that fine with you?

Comment (gartung) [reply lib/spack/spack/relocate.py]: OK. This was an attempt to reproduce what conda does.

File: lib/spack/spack/relocate.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/relocate.py:416]: I'll experiment with that. I think if we don't give the `--force-rpath` option then the relocation will be done on RUNPATH.

File: lib/spack/spack/test/relocate.py [hunk 3]
```diff
@@ -194,3 +217,94 @@ def test_normalize_relative_paths(start_path, relative_paths, expected):
...
+    )
+
+    # Check that the RPATHs changed
+    patchelf = spack.util.executable.which('patchelf')
+    output = patchelf('--print-rpath', str(new_binary), output=str)
+    assert output.strip() == '/foo/lib:/usr/lib64'
+
+
+@pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+def test_relocate_elf_binaries_relative_paths(hello_world, tmpdir):
```
Comment (alalazo) [inline lib/spack/spack/test/relocate.py:287]: This test and the one above can probably be factored better (maybe in a later PR). As a side note I think the fixture `hello_world` might be useful to add further relocation tests.

**Review by gartung — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]
@alalazo: this looks great -- I think the fixture will be really useful going forward for other relocation tests.

It would be good to have a follow-on PR that tests the same logic with `otool` on macOS.

**Pull request merged**